core/src/block.rs

```diff
@@ -989,6 +983,141 @@ impl From<&CommittedBlock> for Vec<Event> {
     }
 }
 
+declare_versioned_with_scale!(VersionedCandidateCommittedBlock 1..2, Debug, Clone, iroha_macro::FromVariant, IntoSchema, Serialize);
+
+impl VersionedCandidateCommittedBlock {
+    /// Converts from `&VersionedCandidateCommittedBlock` to V1 reference
+    #[inline]
+    pub const fn as_v1(&self) -> &CandidateCommittedBlock {
+        match self {
+            Self::V1(v1) => v1,
+        }
+    }
+
+    /// Converts from `&mut VersionedCandidateCommittedBlock` to V1 mutable reference
+    #[inline]
+    pub fn as_mut_v1(&mut self) -> &mut CandidateCommittedBlock {
+        match self {
+            Self::V1(v1) => v1,
+        }
+    }
+
+    /// Performs the conversion from `VersionedCandidateCommittedBlock` to V1
+    #[inline]
+    pub fn into_v1(self) -> CandidateCommittedBlock {
+        match self {
+            Self::V1(v1) => v1,
+        }
+    }
+
+    /// Calculate the hash of the current block.
+    /// `VersionedCandidateCommittedBlock` should have the same hash as `VersionedCommittedBlock`.
+    #[inline]
+    pub fn hash(&self) -> HashOf<Self> {
+        self.as_v1().hash().transmute()
+    }
+
+    /// Returns the header of a valid block
+    #[inline]
+    pub const fn header(&self) -> &BlockHeader {
+        &self.as_v1().header
+    }
+
+    /// Return the signatures (as `payload`) that are verified with the `hash` of this block.
+    #[inline]
+    pub fn verified_signatures(&self) -> impl Iterator<Item = &SignatureOf<Self>> {
+        self.as_v1()
+            .verified_signatures()
+            .map(SignatureOf::transmute_ref)
+    }
+
+    /// Revaidate transactions' hashes and produce [`VersionedCommittedBlock`]
+    ///
+    /// # Errors
+    /// - If transactions' hashes doesn't match hashes stored in block header
+    pub fn revalidate_hashes(self) -> Result<VersionedCommittedBlock, eyre::Report> {
```

Comment:
I believe that additional invariant that `CommittedBlock` offers over `ValidSignedBlock` is that it has enough valid signatures? In that case check in this function is not sufficient to produce `CommittedBlock`

Reply:
if you're going to do the check after creating `CommittedBlock` then we don't need this state. I would consider accepting that as well but then you would have to remove the `CommittedBlock` state from the codebase. In this case you would do the appropriate validation of `ValidSignedBlock` depending on the context and where it's used. But I think I would prefer that we keep the `CommittedBlock` state

Reply:
> I believe that additional invariant that `CommittedBlock` offers over `ValidSignedBlock` is that it has enough valid signatures?

Yes, `CommittedBlock` is block accepted by consensus (i.e. have enough transactions). 

> In that case check in this function is not sufficient to produce `CommittedBlock`.

Good point, this goes for `ValidSignedBlock::commit` method too.
Problem is that signatures verification part is logic strongly related to consensus, since we not only counting signatures but also do filtering by role.

What we probably can do is to provide function `verify_signatures<T>(hash: HashOf<T>) -> bool` inside `revalidate_hashes`. This way logic of signatures verification would still be on `sumeragi` side.

Reply:
I think what would be best is to give all required arguments to the constructor method of `CommittedBlock`. Is it possible to maybe send reference to `sumeragi` or topology to the block constructor? If not, then I guess this sounds like the 2nd best approach, but you would have to write in the comment of the function all the invariants that need to be upheld when calling the constructor. I dislike the 2nd approach because it's not enforced by the type system but at the same time we cannot mark the function `unsafe`

from the top of my head, a 3rd possibility might be to create some intermediate struct in sumeragi that has some invariants and is then used in the block constructor. Not sure if this is viable

Reply:
Problem with 1st approach is that we hardcoded logic of verification.
And sometimes we need to handle logic a bit different: 
1) With genesis block we want to obtain `CommittedBlock` but we don't bothered by amount of signatures. 
2) We have quite a few places where we use `commit` method inside tests and add `Topology` everywhere would be quite troublesome.